go/proxyd/methods.go

```diff
@@ -5,129 +5,247 @@ import (
+		if err != nil {
+			return err
+		}
+		blockNum, err := decodeBlockInput(blockInput)
+		if err != nil {
+			return err
+		}
+		if curBlock <= blockNum+numBlockConfirmations {
+			return nil
+		}
 	}
-	return curBlock <= blockNum+numBlockConfirmations, nil
+
+	key := e.cacheKey(req)
+	return putImmutableRPCResponse(ctx, e.cache, key, req, res)
 }
 
-type EthGetBlockRangeMethod struct {
+type EthGetBlockRangeMethodHandler struct {
+	cache               Cache
 	getLatestBlockNumFn GetLatestBlockNumFn
 }
 
-func (e *EthGetBlockRangeMethod) CacheKey(req *RPCReq) string {
+func (e *EthGetBlockRangeMethodHandler) cacheKey(req *RPCReq) string {
 	start, end, includeTx, err := decodeGetBlockRangeParams(req.Params)
 	if err != nil {
 		return ""
 	}
 	return fmt.Sprintf("method:eth_getBlockRange:%s:%s:%t", start, end, includeTx)
 }
 
-func (e *EthGetBlockRangeMethod) IsCacheable(req *RPCReq) (bool, error) {
+func (e *EthGetBlockRangeMethodHandler) cacheable(req *RPCReq) (bool, error) {
 	start, end, _, err := decodeGetBlockRangeParams(req.Params)
 	if err != nil {
 		return false, err
 	}
 	return !isBlockDependentParam(start) && !isBlockDependentParam(end), nil
 }
 
-func (e *EthGetBlockRangeMethod) RequiresUnconfirmedBlocks(ctx context.Context, req *RPCReq) (bool, error) {
-	curBlock, err := e.getLatestBlockNumFn(ctx)
-	if err != nil {
-		return false, err
+func (e *EthGetBlockRangeMethodHandler) GetRPCMethod(ctx context.Context, req *RPCReq) (*RPCRes, error) {
+	if ok, err := e.cacheable(req); !ok || err != nil {
+		return nil, err
+	}
+
+	key := e.cacheKey(req)
+	return getImmutableRPCResponse(ctx, e.cache, key, req)
+}
+
+func (e *EthGetBlockRangeMethodHandler) PutRPCMethod(ctx context.Context, req *RPCReq, res *RPCRes, blockNumberSync uint64) error {
+	if ok, err := e.cacheable(req); !ok || err != nil {
+		return err
 	}
 
 	start, end, _, err := decodeGetBlockRangeParams(req.Params)
 	if err != nil {
-		return false, err
-	}
-	if isBlockDependentParam(start) || isBlockDependentParam(end) {
-		return true, nil
+		return err
 	}
-	if start == "earliest" && end == "earliest" {
-		return false, nil
+	curBlock, err := e.getLatestBlockNumFn(ctx)
+	if err != nil {
+		return err
 	}
-
 	if start != "earliest" {
 		startNum, err := decodeBlockInput(start)
 		if err != nil {
-			return false, err
+			return err
 		}
 		if curBlock <= startNum+numBlockConfirmations {
-			return true, nil
+			return nil
 		}
 	}
 	if end != "earliest" {
 		endNum, err := decodeBlockInput(end)
 		if err != nil {
-			return false, err
+			return err
 		}
 		if curBlock <= endNum+numBlockConfirmations {
-			return true, nil
+			return nil
 		}
 	}
-	return false, nil
+
+	key := e.cacheKey(req)
+	return putImmutableRPCResponse(ctx, e.cache, key, req, res)
+}
+
+type EthCallMethodHandler struct {
+	cache               Cache
+	getLatestBlockNumFn GetLatestBlockNumFn
+}
+
+func (e *EthCallMethodHandler) cacheKey(req *RPCReq) string {
+	type ethCallParams struct {
```

Comment:
You need a blocktag here as well because the same `eth_call` can return different values at different heights. Context specific blocktags such as `latest` and `pending` cannot be ever cached, and those are the most common

Reply:
Because of this, I don't particularly love the idea of caching `eth_call` - its really only good for archival access

Reply:
https://docs.ethers.io/v5/api/providers/types/#providers-BlockTag

Reply:
good catch! Need to fix that.

Reply:
well actually, I don't :-). I mean the `eth_call` cache doesn't rely on the blocktag input. But rather callers are expected to provide an explicit block number that is synchronized with the RPCRes being cached.

Reply:
In the event that `eth_call` is used for the `latest` block, we resolve `latest` to the current block number for the `PutRPC` call - see rationale [here](https://github.com/ethereum-optimism/optimism/pull/2029/files/61810cc5491056891951f98ec6548a5ccdf00c9b#diff-72c3fe5d29ecb58f988d4e893921ab3f18b4858c7cab8322fb0551d8692a70fdR89)

Reply:
It feels like resolving `latest` could be subject to race conditions?